### PR TITLE
[FSDP()][3/N] Refactor public APIs

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87937 [FSDP] Simplify `_reset_lazy_init()`
* #87936 [FSDP()][22/N] Refactor `_cast_buffers()` in `_lazy_init()`
* #87935 [FSDP()][21/N] Refactor `_buffer_name_to_orig_dtype` computation
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward
* #87926 [FSDP()][13/N] Refactor unshard/reshard/grads
* #87925 [FSDP()][12/N] Easy cleanup
* #87924 [FSDP()][10/N][11/N] Introduce composable (ctor only)
* #87923 [FSDP()][9/N] Refactor ctor (continued)
* #87922 [FSDP()][8/N] Refactor limiter's `_FreeEventQueue`
* #87921 [FSDP()][7/N] Refactor most of ctor
* **#87917 [FSDP()][3/N] Refactor public APIs**

- This PR defines a new `api.py` meant to hold the public API for FSDP (minus `FullyShardedDataParallel` itself). This is needed because several of the `_<...>_utils.py` files rely on the public API, and we cannot import from `torch.distributed.fsdp.fully_sharded_data_parallel` without a circular import. Calling the file `api.py` follows the convention used by `ShardedTensor`.
- This PR cleans up the wording in the `BackwardPrefetch`, `ShardingStrategy`, `MixedPrecision`, and `CPUOffload` docstrings.
- This PR adds the aforementioned classes to `fsdp.rst` to have them rendered in public docs.
- To abide by the public bindings contract (`test_public_bindings.py`), the aforementioned classes are removed from `fully_sharded_data_parallel.py`'s `__all__`. This is technically BC breaking if someone uses `from torch.distributed.fsdp.fully_sharded_data_parallel import *`; however, that does not happen in any of our own external or internal code.

### CODE REVIEW

File: torch/distributed/fsdp/common_utils.py
Comment (mrshenli) [inline torch/distributed/fsdp/common_utils.py:9]: does this suppose to show in our API doc?

**Review by mrshenli — APPROVED** [review]
failure is real, please address before landing. 

```
WARNING: autodoc: failed to import module 'fsdp' from module 'torch.distributed'; the following exception was raised:
WARNING: autodoc: failed to import class 'fsdp.FullyShardedDataParallel' from module 'torch.distributed'; the following exception was raised:
```

code changes LGTM. If the the plan is to display public common utils (I assume this is the case, as users will see `BackwardPrefetch`), please also make sure the docs are rendered correctly. 

Stamp to unblock.

File: torch/distributed/fsdp/common_utils.py
Comment (mrshenli) [inline torch/distributed/fsdp/common_utils.py:29]: This paragraph seems should be either a TODO comment for us (i.e., not in docs), or should be a warning with explicit work-around suggestions.

File: torch/distributed/fsdp/common_utils.py
Comment (mrshenli) [inline torch/distributed/fsdp/common_utils.py:26]: would I be correct, if we are not doing prefetch, we should have even lower memory footprint, because reduce-scatter would have deleted full grad before allgather?

File: torch/distributed/fsdp/common_utils.py
Comment (mrshenli) [inline torch/distributed/fsdp/common_utils.py:27]: For future PR, we might also want to point users to rate limiter here just in case they do run into memory issues after turning on prefetching.

File: torch/distributed/fsdp/common_utils.py
Comment (awgu) [reply torch/distributed/fsdp/common_utils.py:26]: This is a good question, and it has made me take a step back about the `BACKWARD_PRE` docs too.

TL;DR: I think we should simplify the docs to say that both prefetching modes may increase peak memory and omit the explanation since it might be too confusing.

> if we are not doing prefetch, we should have even lower memory footprint, because reduce-scatter would have deleted full grad before allgather?

I am not sure if this is true. Even without prefetching, if FSDP freed the unsharded gradient, the next all-gather cannot reuse that block since it is from a different stream. Whether the next gradient computation can reuse the memory from the unsharded gradient depends on the CPU thread timing.

File: torch/distributed/fsdp/common_utils.py
Comment (awgu) [reply torch/distributed/fsdp/common_utils.py:9]: Yes. I will make sure the docs are rendering correctly before landing.

**Pull request closed without merging**
